Bitcoin transactiona rewind

gabr1gus opened the pull request:
Implementado um prototipo da função de rewind
Adicionado uma maneira de impedir o processamento de blocos se a rede nao esta sincronizada

gabr1gus committed:
prototipo de funçao de rewind 

--não testado
--prototipo da funcionalidade de bloquear processamento de bloco se nao sincronizado

gabr1gus committed:
Merge remote-tracking branch 'origin/master' into bitcoin-transactions-rewind

gabr1gus committed:
RPC use lowered 

--testado para influxo de blocos parados por 24 horas
--aparentemente esta enviando transações ja existentes no banco de dados, possível erro de corrida
--pequenas correções na formatação

gabr1gus committed:
minor changes

--percebi que não tinha mandado isso ainda

gabr1gus committed:
trocado a forma de verificar sincronizaçao da blockchain

--agora o processo é feito por RPC
--processo antigo (get para api externa) foi deixado como failsafe

GCastilho commented:
Em que situação um comando RPC falha mas o sistema ainda está sincronizando pra precisar desse fail-safe?

Não seria mais sensato simplesmente tentar o rpc de novo?

gabr1gus commented:
> Em que situação um comando RPC falha mas o sistema ainda está sincronizando pra precisar desse fail-safe?
> 
> Não seria mais sensato simplesmente tentar o rpc de novo?

tem razão, vou simplesmente jogar pra error e fechar o sistema antes mesmo dele ser iniciado direito, no fim isso nao faz muita diferença

gabr1gus committed:
remoçao do failsafe desnecessario

GCastilho commented:
>vou simplesmente jogar pra error e fechar o sistema antes mesmo dele ser iniciado direito

Não precisa. Sério, pq vc não simplesmente faz ele ficar tentando a cada 10s?

Coloca uma mensagem de erro na tela (da primeira vez) e tenta de novo

Se fizer assim eu não posso ter o módulo rodando sem o BTC rodando, que é uma coisa q hoje é possível e não tem nada de errado

gabr1gus committed:
Merge remote-tracking branch 'origin/master' into bitcoin-transactions-rewind

gabr1gus committed:
adicionado retry ao tentar adquirir o block height

--alterado abstract initBlockchainListener() para async
--alterado concrete initBlockchainListener() da nano ee bitcin para async

GCastilho commented:
Falta implementar o que, @gabr1gus ?

gabr1gus committed:
testado e feito pequenas correçoes

--adicionado um wait que funciona (30s de wait)
--pequenas alteraçoes gerais
--corrigido identaçao
--corrigido chamada errada do rpc
--minor changes

gabr1gus committed:
Merge branch 'master' into bitcoin-transactions-rewind

gabr1gus commented:
eu coloquei um wait de 30 segundos para o check, as mensagens de "nao possui blockheight ainda" provavelmente sao desncessarias, nao sei se é uma boa ideia remover os console log ou nao

GCastilho commented on external_modules/bitcoin/index.ts:
```diff
@@ -10,16 +10,19 @@ export class Bitcoin extends Common {
 	mainServerIp = MAIN_SERVER_IP
 	mainServerPort = MAIN_SERVER_PORT
-
+	blockHeight = 20000000 //inicializado com um valor extremamente alto
```
Por que precisa começar com um valor alto?

gabr1gus replied:
para nao passar num teste de  newBlock.blockheight >= blockHeight se o blockheight nao tiver sido recuperado ainda, numa situaçao ideal, esse valor fica inicializado por pouco tempo
sem esse teste quando vier uma enxurrada de blocos, numa sincronização, o rpc não aguenta
provavelmente tem uma forma melhor de fazer isso, mas eu só estava procurando uma forma que funcionasse na epoca, e nao consigo pensar em uma forma melhor atualmente

GCastilho commented on external_modules/bitcoin/index.ts:
```diff
@@ -10,16 +10,19 @@ export class Bitcoin extends Common {
 	mainServerPort = MAIN_SERVER_PORT
-
+	blockHeight = 20000000 //inicializado com um valor extremamente alto
+	canSincronize = true
```
O que essa variável faz?

gabr1gus replied:
impede que comece a sincronizar enquanto o bitcoin nao encontrou o utimo bloco (block height seria a altura atual do bloco que esta mais sincronizado na rede), a variavel canSincronize eu nao lembro, vou dar uma olhada daqui a pouco mas acho que ela servia pra dizer que pode ou nao sincronizar pois esta sincronizando nesse momento

GCastilho commented on external_modules/bitcoin/index.ts:
```diff
@@ -39,11 +42,36 @@ export class Bitcoin extends Common {
+			return new Promise(resolve => setTimeout(resolve, time))
+		}
+		let blockHeight = null
+		const teste = await methods.rpc.getBlockCount()
+		console.log('\n\n\n' + teste + '\n\n\n')
```
É que remover esses testes, certo?

gabr1gus replied:
esses em especificos nao deveriam nem estar nesse comit(ja removo eles, só coloquei pra ver se o rpc tava funcionando), eu me referia aos console log que tem caso tenha um erro e nao consiga encontrar o blockHeight (rpc error normalmente)

GCastilho replied:
Então **esse** console.log tem que tirar, mas um outro tem que deixar, é isso?

GCastilho commented on external_modules/bitcoin/index.ts:
```diff
@@ -39,11 +42,36 @@ export class Bitcoin extends Common {
+				blockHeight = (await this.rpc.getBlockChainInfo())?.headers
+			} catch (e) {
+				console.log('failed to recover block height')
+				if (e.name != 'RpcError' && e.code != 'ECONNREFUSED')
+					console.log(e)
```
Pode explicar o que esse if está checando?

gabr1gus replied:
rpc error pode acontecer quando o bitcoin esta abrindo e enconrefused quando o bitcoin nao esta nem aberto, sao erros "esperados" provavelmente eu estou ignorando mais erros que o necessario com esse check, mas eu nao me importo com que erro de, se o bitcoin nao mudar aleatoriamente o comando rpc ou algo do genero, eventualmente o sistema vai sair do loop eventualmente,só que se der um erro inesperado, tem que ver o que é

GCastilho commented on external_modules/bitcoin/index.ts:
```diff
@@ -39,11 +42,36 @@ export class Bitcoin extends Common {
+				console.log('retring....')
+			}
+			console.log('the block is not updated')
+			await sleep(30000)//30 segundos
```
```suggestion
			await sleep(30000) //30 segundos
```

GCastilho commented on external_modules/bitcoin/index.ts:
```diff
@@ -39,11 +42,36 @@ export class Bitcoin extends Common {
+
+		this.blockHeight = blockHeight
+
+		console.log('current block height :' + this.blockHeight)
```
```suggestion
		console.log('Current block height:' + this.blockHeight)
```

GCastilho commented on external_modules/bitcoin/methods/processBlock.ts:
```diff
@@ -75,7 +75,10 @@ export function processBlock(this: Bitcoin) {
 	const _processBlock = async (block: string) => {
 		if (typeof block != 'string') return
+		const blockCount = await this.rpc.getBlockCount()
+		if (blockCount < this.blockHeight || !this.canSincronize) return
```
Por que ele retorna se o this.blockHeight for maior que o blockCount?

GCastilho commented on external_modules/bitcoin/methods/rewindTransactions.ts:
```diff
@@ -0,0 +1,107 @@
+
+
+export async function rewindTransactions(this: Bitcoin, newBlockhash: string) {
+	this.canSincronize = false
```
Por que?

GCastilho commented on external_modules/bitcoin/methods/rewindTransactions.ts:
```diff
@@ -0,0 +1,107 @@
+
+	if (!blockhash) {
+		blockhash = (await this.rpc.blockInfo(newBlockhash))?.previousblockhash
+		if (!blockhash)
+			return
```
```suggestion
		if (!blockhash) return
```

GCastilho commented on external_modules/bitcoin/methods/rewindTransactions.ts:
```diff
@@ -0,0 +1,107 @@
+			upsert: true
+		})
+	}
+	const { transactions }: any = await this.rpc.listSinceBlock(blockhash)
```
```suggestion
	const { transactions } = await this.rpc.listSinceBlock(blockhash)
```

GCastilho commented on external_modules/bitcoin/methods/rewindTransactions.ts:
```diff
@@ -0,0 +1,107 @@
+
+	for (const transaction of transactions) {
+		try {
+			const _transaction = await formatTransaction(transaction)
```
```suggestion
			const formattedTransaction = await formatTransaction(transaction)
```

GCastilho commented on external_modules/bitcoin/methods/rewindTransactions.ts:
```diff
@@ -0,0 +1,107 @@
+		try {
+			const _transaction = await formatTransaction(transaction)
+			if (!_transaction) return
+			console.log( 'received transaction', _transaction ) //remove
```
```suggestion
			console.log('received new transaction:', formattedTransaction )
```

GCastilho commented on external_modules/bitcoin/methods/rewindTransactions.ts:
```diff
@@ -0,0 +1,107 @@
+				txid : _transaction.txid,
+				type: 'receive',
+				account: _transaction.account
+			}).save().catch()
```
Esse catch está sem um handler

gabr1gus replied:
alguma ideia de que tipo de handler eu colocaria nisso ? se der duplicate key, nao é um erro que deveria ser tratado

GCastilho replied:
Faz ele ignorar o duplicate key e se for outro dá erro msm, pq daí teve mesmo alguma coisa errada

GCastilho commented on external_modules/bitcoin/methods/rewindTransactions.ts:
```diff
@@ -0,0 +1,107 @@
+		}
+	}
+	await meta.updateOne({
+		info: 'lastSyncBlock'
```
Pq salva esse info?

gabr1gus replied:
se me lembro bem, era para eu pegar todas as transações feitas depois desse bloco, é necessário salvar ele em algum lugar

GCastilho replied:
mas vc tá salvando uma string nesse info

gabr1gus replied:
se bem me lembro, a ideia era que eu pudesse reutilizar o modelo se outras currencys precisassem de dados pra fazer alguma coisa

GCastilho replied:
Eu ainda n entendi. Pq tá salvando a string 'lastSyncBlock' aqui?

gabr1gus replied:
é o "nome" da info que esta sendo salva, serve para eu encontrar depois qual informação eu quero por esse nome, o que importa mesmo é o que esta sendo salvo em details

GCastilho commented on external_modules/common/db/models/meta.ts:
```diff
@@ -0,0 +1,22 @@
+	details: string
+}
+
+const Meta: Schema = new Schema({
```
```suggestion
const Schema = new Schema({
```

gabr1gus replied:
deve funcionar, vou dar commit e testar mais tarde

GCastilho commented on external_modules/nano/index.ts:
```diff
@@ -23,7 +23,7 @@ export class Nano extends Common {
 	findMissingTx = methods.findMissingTx
 
-	initBlockchainListener = methods.nanoWebSocket
+	initBlockchainListener = async () => {methods.nanoWebSocket}
```
Por que?

gabr1gus replied:
eu não lembro

GCastilho replied:
nice (descubra pfvr)

GCastilho reviewed (changes requested):
Algumas perguntas e sugestões, mas parece ok

gabr1gus committed:
Update external_modules/bitcoin/index.ts

--GCastilho sugestion

Co-authored-by: GCastilho <34466758+GCastilho@users.noreply.github.com>

gabr1gus committed:
Update external_modules/common/db/models/meta.ts

--GCastilho sugestion

Co-authored-by: GCastilho <34466758+GCastilho@users.noreply.github.com>

GCastilho committed:
Corrigido um bug com o nome de um schema

-- Corrigido o nome do schema meta que estava conflitando com outra variável
-- Adicionado ts-node-dev e o typescript aos módulos externos

gabr1gus committed:
Update index.ts

--removido log que dava erro se a bitcoin nao tivesse aberta

GCastilho committed:
Refatorado o log de erros

-- Refatorado o log de erros do block height
-- Refatorado outras coisas pequenas

GCastilho committed:
Small fix

gabr1gus committed:
Merge branch 'master' into bitcoin-transactions-rewind

gabr1gus committed:
minor changes

--modificado o bloco de sincronizaçao inicial
--renomeado canSincronize para Synchronizing
--minor changes

gabr1gus committed:
minor fix

--mudado blockhash do bloco -1 para o bloco 0
--renomeado synchronizing para rewinding
--minor changes

gabr1gus committed:
--minor fix

GCastilho committed:
Refatorado umas paradas

-- Refatorado umas parte dos códigos e pá com o @gabr1gus
-- Minor changes

GCastilho committed:
Meta update

GCastilho committed:
Corrigido um bug onde apenas uma transação era salva

-- Corrigido um bug no index do mongo que está impedindo mais de uma transação sem opid de ser salva ao mesmo tempo

GCastilho committed:
Update bitcoin.conf

Removido config duplicada

GCastilho reviewed (dismissed):
An waaaay we go

GCastilho committed:
Resolvido retorno do initBlockchainListener precisar retornar uma promessa

-- O initBlockchainListener agora pode retornar uma promessa com void OU void
